### PR TITLE
581 add intent delivery failed error and document errors

### LINKED ISSUE #581: Add an error to ResolveError for failure to deliver an intent to an app
## Enhancement Request
Add a `IntentDeliveryFailed` error to the `ResolveError` enumeration.

### Use Case:
The [ResolveError](https://fdc3.finos.org/docs/next/api/ref/Errors#resolveerror) enumeration provides standard string for failures you might get back when raising an intent or `findIntent` runs into a problem (e.g. the are no apps available for that intent, the resolver isn't available or the user cancelled resolution).  However, it doesn't include a standardized error for cases where it succeeded in resolving to an app or app instance, but couldn't deliver the intent and context to it. This might occur because: 

1. the app failed to add an intent listener or, 
2. added an intent listener but not for the right type
3. or failed in some other way (e.g. it closed immediately).

### Additional Information
There is some overlap with the errors added in the recently merged PR for targetting app instances:
- #509 
- `TargetAppUnavailable`
- `TargetInstanceUnavailable`

However, these are intended to indicate that a specifically targetted app or instance is not available to be started or has been closed (e.g. option for an instance was shown in a resolver, but it closed before delivery or the target app couldn't be started for some reason). This handles case 3 above, but not 1 or 2.

Clear documentation of which error to use when will be added to ensure the various situations are easily differentiated.

### PR DESCRIPTION
**Pull request by kriswest**
resolves #581 by adding a new IntentDeliveryFailed error
Also:
- Adds documentation of the intended use of each error
- Removes unnecessary copyright dates from source files (the years is not being updated when changes are made and the github repo already timestamps when content was added in its revision history)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -13,11 +13,12 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
...
 * Added the ability to return data from an intent, via the addition of an IntentHandler type and a `getResult()` to IntentResolution, both of which return a Promise of a Context object. ([#495](https://github.com/finos/FDC3/pull/495))
 * Added error `UserCancelled` to the `ResolveError` enumeration to be used when user closes the resolver UI or otherwise cancels resolution of a raised intent ([#522 ](https://github.com/finos/FDC3/pull/522))
-* Added an `instanceId` (and optional `instanceMetadata`) field to `AppMetadata` allowing it to refer to specific app instances and thereby supporting targetting of intents to specific app instances. Also added a `findInstanes()` function to the desktop agent. ([#509]((https://github.com/finos/FDC3/pull/509))
+* Added an `instanceId` (and optional `instanceMetadata`) field to `AppMetadata` allowing it to refer to specific app instances and thereby supporting targetting of intents to specific app instances. Also added a `findInstanes()` function to the desktop agent and `TargetAppUnavailable` and `TargetInstanceUnavailable` Errors to the `ResolveError` enumeration. ([#509]((https://github.com/finos/FDC3/pull/509))
```
Comment (mattjamieson) [inline CHANGELOG.md:16]:
```suggestion
* Added an `instanceId` (and optional `instanceMetadata`) field to `AppMetadata` allowing it to refer to specific app instances and thereby supporting targetting of intents to specific app instances. Also added a `findInstances()` function to the desktop agent and `TargetAppUnavailable` and `TargetInstanceUnavailable` Errors to the `ResolveError` enumeration. ([#509]((https://github.com/finos/FDC3/pull/509))
```

File: docs/api/ref/Errors.md [hunk 2]
```diff
@@ -61,9 +75,12 @@ Contains constants representing the errors that can be encountered when calling
...
 ```typescript
 enum ChannelError {
-  NoChannelFound = "NoChannelFound",
-  AccessDenied = "AccessDenied",
-  CreationFailed = "CreationFailed"
+  /** Returned if the specified channel is not found when attempting to join a channel via the `joinUserChannel` function  of the DesktopAgent (`fdc3`).*/
+  NoChannelFound = 'NoChannelFound',
+  /** SHOULD be returned when a request to join a user channel or to a retrieve a Channel object via the `joinUserChannel` or `getOrCreateChannel` methods of the DesktopAgent (`fdc3`) object is denied. */
+  AccessDenied = 'AccessDenied',
+  /** SHOULD Returned when a channel cannot be created or retrieved via the `getOrCreateChannel` method of the DesktopAgent (`fdc3`).*/
```
Comment (mattjamieson) [inline docs/api/ref/Errors.md:82]:
```suggestion
  /** SHOULD be returned when a channel cannot be created or retrieved via the `getOrCreateChannel` method of the DesktopAgent (`fdc3`).*/
```

File: src/api/Errors.ts [hunk 3]
```diff
@@ -1,31 +1,50 @@
...
+  /** Returned if the Intent handler function processing the raised intent throws an error or rejects the promise it returned. */
   IntentHandlerRejected = 'IntentHandlerRejected',
 }
 
 export enum ChannelError {
+  /** Returned if the specified channel is not found when attempting to join a channel via the `joinUserChannel` function  of the DesktopAgent (`fdc3`).*/
   NoChannelFound = 'NoChannelFound',
+  /** SHOULD be returned when a request to join a user channel or to a retrieve a Channel object via the `joinUserChannel` or `getOrCreateChannel` methods of the DesktopAgent (`fdc3`) object is denied. */
   AccessDenied = 'AccessDenied',
+  /** SHOULD Returned when a channel cannot be created or retrieved via the `getOrCreateChannel` method of the DesktopAgent (`fdc3`).*/
```
Comment (mattjamieson) [inline src/api/Errors.ts:48]:
```suggestion
  /** SHOULD be returned when a channel cannot be created or retrieved via the `getOrCreateChannel` method of the DesktopAgent (`fdc3`).*/
```

**Review by mattjamieson — CHANGES_REQUESTED** [review]
Looks good to me. Added suggestions to fix a few typos.

**Comment by kriswest** [comment]
Thanks @mattjamieson, suggestions applied

**Review by mattjamieson — APPROVED** [review]

File: docs/api/ref/Errors.md [hunk 4]
```diff
@@ -24,12 +28,20 @@ Contains constants representing the errors that can be encountered when calling
...
 export enum ResolveError {
+  /** SHOULD be returned if no apps are available that can resolve the intent and context combination.*/
   NoAppsFound = 'NoAppsFound',
+  /** Returned if the FDC3 desktop agent implementation is not currently able to handle the request.*/
   ResolverUnavailable = 'ResolverUnavailable',
+  /** Returned if the user cancelled the resolution request, for example by closing or cancelling a resolver UI.*/
   UserCancelled = 'UserCancelledResolution',
+  /** SHOULD be returned if a timeout cancels intent resolution, that required user interaction. Please used `ResolverUnavailable` instead for situations where a resolver UI or similar fails.*/
   ResolverTimeout = 'ResolverTimeout',
+  /** Returned if a specified target application is not available or new instance of it cannot be opened. */
```
Comment (greyseer256) [inline docs/api/ref/Errors.md:39]:
```suggestion
  /** Returned if a specified target application is not available or a new instance of it cannot be opened. */
```

File: docs/api/ref/Errors.md [hunk 5]
```diff
@@ -45,7 +57,9 @@ Contains constants representing the errors that can be encountered when calling
 
 ```typescript
 enum DataError {
+  /** Returned if the intent handler exited without returning Promise or that Promise was not resolved with a Context object. */
```
Comment (greyseer256) [inline docs/api/ref/Errors.md:60]:
```suggestion
  /** Returned if the intent handler exited without returning a Promise or that Promise was not resolved with a Context object. */
```

File: docs/api/ref/Errors.md [hunk 6]
```diff
@@ -45,7 +57,9 @@ Contains constants representing the errors that can be encountered when calling
...
   NoDataReturned = 'NoDataReturned',
+  /** Returned if the Intent handler function processing the raised intent throws an error or rejects the promise it returned. */
```
Comment (greyseer256) [inline docs/api/ref/Errors.md:62]: inconsistent capitalization of 'promise'

File: docs/api/ref/Errors.md [hunk 7]
```diff
@@ -61,9 +75,12 @@ Contains constants representing the errors that can be encountered when calling
 
 ```typescript
 enum ChannelError {
-  NoChannelFound = "NoChannelFound",
-  AccessDenied = "AccessDenied",
-  CreationFailed = "CreationFailed"
+  /** Returned if the specified channel is not found when attempting to join a channel via the `joinUserChannel` function  of the DesktopAgent (`fdc3`).*/
```
Comment (greyseer256) [inline docs/api/ref/Errors.md:78]:
```suggestion
  /** Returned if the specified channel is not found when attempting to join a channel via the `joinUserChannel` function of the DesktopAgent (`fdc3`).*/
```
(double space)

File: src/api/Errors.ts [hunk 8]
```diff
@@ -1,31 +1,50 @@
...
 export enum ResolveError {
+  /** SHOULD be returned if no apps are available that can resolve the intent and context combination.*/
   NoAppsFound = 'NoAppsFound',
+  /** Returned if the FDC3 desktop agent implementation is not currently able to handle the request.*/
   ResolverUnavailable = 'ResolverUnavailable',
+  /** Returned if the user cancelled the resolution request, for example by closing or cancelling a resolver UI.*/
   UserCancelled = 'UserCancelledResolution',
+  /** SHOULD be returned if a timeout cancels intent resolution, that required user interaction. Please used `ResolverUnavailable` instead for situations where a resolver UI or similar fails.*/
   ResolverTimeout = 'ResolverTimeout',
+  /** Returned if a specified target application is not available or new instance of it cannot be opened. */
```
Comment (greyseer256) [inline src/api/Errors.ts:28]:
```suggestion
  /** Returned if a specified target application is not available or a new instance of it cannot be opened. */
```

File: src/api/Errors.ts [hunk 9]
```diff
@@ -1,31 +1,50 @@
...
   TargetAppUnavailable = 'TargetAppUnavailable',
+  /** Returned if a specified target application instance is not available, for example because it has been closed. */
   TargetInstanceUnavailable = 'TargetInstanceUnavailable',
+  /** Returned if the intent and context could not be delivered to the selected application or instance, for example because it has not added an intent handler within a timeout.*/
+  IntentDeliveryFailed = 'IntentDeliveryFailed',
 }
 
 export enum DataError {
+  /** Returned if the intent handler exited without returning Promise or that Promise was not resolved with a Context object. */
```
Comment (greyseer256) [inline src/api/Errors.ts:37]:
```suggestion
  /** Returned if the intent handler exited without returning a Promise or that Promise was not resolved with a Context object. */
```

File: src/api/Errors.ts [hunk 10]
```diff
@@ -1,31 +1,50 @@
...
   NoDataReturned = 'NoDataReturned',
+  /** Returned if the Intent handler function processing the raised intent throws an error or rejects the promise it returned. */
```
Comment (greyseer256) [inline src/api/Errors.ts:39]: inconsistent capitalization of 'promise'

File: src/api/Errors.ts [hunk 11]
```diff
@@ -1,31 +1,50 @@
...
 
+/** Constants representing the errors that can be encountered when calling the `findIntent`, `findIntentsByContext`, `raiseIntent` or `raiseIntentForContext` methods on the DesktopAgent (`fdc3`). */
 export enum ResolveError {
+  /** SHOULD be returned if no apps are available that can resolve the intent and context combination.*/
   NoAppsFound = 'NoAppsFound',
+  /** Returned if the FDC3 desktop agent implementation is not currently able to handle the request.*/
   ResolverUnavailable = 'ResolverUnavailable',
+  /** Returned if the user cancelled the resolution request, for example by closing or cancelling a resolver UI.*/
   UserCancelled = 'UserCancelledResolution',
+  /** SHOULD be returned if a timeout cancels intent resolution, that required user interaction. Please used `ResolverUnavailable` instead for situations where a resolver UI or similar fails.*/
```
Comment (greyseer256) [inline src/api/Errors.ts:26]:
```suggestion
  /** SHOULD be returned if a timeout cancels an intent resolution that required user interaction. Please use `ResolverUnavailable` instead for situations where a resolver UI or similar fails.*/
```

File: src/api/Errors.ts (hunk 10 above)
Comment (kriswest) [inline src/api/Errors.ts:39]:
```suggestion
  /** Returned if the Intent handler function processing the raised intent throws an error or rejects the Promise it returned. */
```

Comment (kriswest) [reply src/api/Errors.ts]: fixed

**Comment by kriswest** [comment]
Suggestions applied @greyseer256, thanks again

**Review by greyseer256 — APPROVED** [review]

File: src/api/Errors.ts [hunk 12]
```diff
@@ -1,31 +1,50 @@
...
+  /** Returned if a specified target application is not available or a new instance of it cannot be opened. */
   TargetAppUnavailable = 'TargetAppUnavailable',
+  /** Returned if a specified target application instance is not available, for example because it has been closed. */
   TargetInstanceUnavailable = 'TargetInstanceUnavailable',
+  /** Returned if the intent and context could not be delivered to the selected application or instance, for example because it has not added an intent handler within a timeout.*/
+  IntentDeliveryFailed = 'IntentDeliveryFailed',
 }
 
 export enum ResultError {
+/** Returned if the intent handler exited without returning a Promise or that Promise was not resolved with a Context or Channel object. */
```
Comment (kriswest) [inline src/api/Errors.ts:37]:
```suggestion
  /** Returned if the intent handler exited without returning a Promise or that Promise was not resolved with a Context or Channel object. */
```

**Pull request merged**